cpp/tests/test_metadata_payload_exchange.cpp

```diff
@@ -0,0 +1,370 @@
+            peer_rank, std::move(metadata), std::move(data_buffer)
+        );
+    }
+
+    std::unique_ptr<Buffer> allocate_receive_buffer(std::size_t size) {
+        return br->allocate(stream, br->reserve_or_fail(size, MemoryType::DEVICE));
+    }
+
+    void wait_for_communication_complete() {
+        for (int iter = 0; iter < 100 && !comm_interface->is_idle(); ++iter) {
+            comm_interface->recv();
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        }
+    }
+
+    Communicator* comm;
+    std::unique_ptr<rmm::mr::device_memory_resource> mr;
+    rmm::cuda_stream_view stream;
+    std::unique_ptr<BufferResource> br;
+    std::shared_ptr<Statistics> statistics;
+    std::unique_ptr<TagMetadataPayloadExchange> comm_interface;
+};
+
+TEST_F(MetadataPayloadExchangeTest, InitialState) {
+    // Communication interface should start in idle state
+    EXPECT_TRUE(comm_interface->is_idle());
+}
+
+TEST_F(MetadataPayloadExchangeTest, SetDataOnAlreadySetData) {
+    // Test that set_data throws when data is already set
+    std::vector<std::uint8_t> test_metadata = {0x01};
+    constexpr std::size_t data_size = 256;
+
+    // Create a message with data already set
+    auto message = create_test_message(Rank{0}, test_metadata, data_size);
+
+    // Try to set data again - should throw
+    auto buffer = allocate_receive_buffer(128);
+    EXPECT_THROW(message->set_data(std::move(buffer)), std::logic_error);
+}
+
+TEST_F(MetadataPayloadExchangeTest, SendReceiveMetadataOnly) {
+    if (comm->nranks() < 2) {
+        GTEST_SKIP() << "Test requires at least 2 ranks";
+    }
+
+    Rank peer_rank = (comm->rank() + 1) % comm->nranks();
+    std::vector<std::uint8_t> test_metadata = {0x01, 0x02, 0x03, 0x04};
+
+    if (comm->rank() == 0) {
+        // Rank 0 sends metadata-only message
+        std::vector<std::unique_ptr<MetadataPayloadExchange::Message>> messages;
+        messages.push_back(create_test_message(peer_rank, test_metadata));
+
+        EXPECT_TRUE(comm_interface->is_idle());
+        comm_interface->send(std::move(messages));
+        EXPECT_FALSE(comm_interface->is_idle());
+    }
+
+    std::vector<std::unique_ptr<MetadataPayloadExchange::Message>> received_messages;
+    for (int iter = 0; iter < 10 && received_messages.empty(); ++iter) {
+        auto messages = comm_interface->recv();
+        received_messages.insert(
+            received_messages.end(),
+            std::make_move_iterator(messages.begin()),
+            std::make_move_iterator(messages.end())
+        );
+
+        if (!received_messages.empty())
+            break;
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    }
+
+    if (comm->rank() == peer_rank) {
+        EXPECT_EQ(received_messages.size(), 1);
+        auto& msg = received_messages[0];
+        EXPECT_EQ(msg->peer_rank(), 0);
+        EXPECT_EQ(msg->metadata(), test_metadata);
+        EXPECT_EQ(msg->data(), nullptr);
+    }
```

Comment:
This never fires, right? Because `peer_rank` is defined as `(comm->rank() + 1) % nranks` so it is never true that `comm->rank() == peer_rank`.